### ISSUE TITLE
Custom codec not used for decoding

### ISSUE THREAD

**Issue by Vinarnt** [MorphiaOrg/morphia#1432]
```
morphia: 2.0.0-BETA-1/SNAPSHOT
JDK: openjdk 13.0.2 2020-01-14
mongodb: 4.2.3
```

Hi,

I'm in the process of migrating an app from morphia 1.3.2 to the new 2.0 version and I'm stuck with a case that might be a bug as it was working previously and should works on the new version.

Basically I have to save a map of mixed objects. Those objects inherit a common class that contains a single value.
You can see the corresponding code for it [here](https://gist.github.com/Vinarnt/174703c74440195b2652bc8e8373a176).

It works great for primitives but for third party objects that need to be transformed on insert and find it's not working as expected.
When I insert/update the `GuildProperties` object with the `properties` member set with a mapped object of type `RolePropertyValue`, the codec for the `Role` object is called to write the id of the role it wrap.
But when fetching it from the database, the codec is not called and thus fallback to a default one leaving me with a `RolePropertyValue` containing a `Long` corresponding to the id of the role previously encoded, instead of getting the `Role` object back. (Don't even know how it is possible)
I tested setting the concrete class of the property value with the Property annotation, but got a NPE caused by a null codec on the model as the `Role` class is not in the discriminator lookup.

The stored schema is like this:
```json
{
    "_id": {
        "$oid": "5e723d712272e97e4eb58aa4"
    },
    "_t": "GuildProperties",
    "guild_id": {
        "$numberLong": "269552852760920068"
    },
    "properties": {
        "NOTICES": {...},
        "AVAILABLE_SELFROLES": {...},
        "AUTOROLE": {
            "_t": "RolePropertyValue",
            "value": {
                "$numberLong": "370194687635619842"
            }
        },
        "COMMAND_PREFIX": {...}
    }
}
```

Is it intended behavior or am I missing something? Struggling for it for a few days, I couldn't find a way to sort this out without making changes to morphia.

**Comment by evanchooly** [comment]
You need to register that codec.  I realized last night that I have updated all the other docs except for the getting started docs.  ( I thought I had but apparently not. )  I'm planning on updating those tonight and pushing them online.  In short, you have to register your custom codec when you create your `MongoClient`.  If you're not manually creating one and letting Morphia do that for you, there is currently no way to do so.  That's an oversight and I'll correct that at the time I'm documenting how to do it.  :)

**Comment by Vinarnt** [comment]
Yes I know it's registered with a CodecProvider, that's why it actualy call the encode method.
But when reading back the object, it doesn't use the codec.

**Comment by Vinarnt** [comment]
I set a testcase for this.
https://github.com/Vinarnt/morphia-generic-map-testcase

When running the test you'll run in an exception, a hint that something is not going as it should.

**Comment by evanchooly** [comment]
From what I'm seeing the codecs are being used just fine.  It just seems you have something funky in your logic and setup that's running amok.  The error I get is:

```
class java.lang.Long cannot be cast to class external.Role
```

And i think this is because you're trying to compare things with fields mapped as generics.  I say this because I can see Morphia loading the `long` value from the database correctly but you're trying to compare to another instance of `RolePropertyValue` that you've created using a `Role` as that field type.  These are obviously incompatible thus the CCE.

In short, there *is* a problem here but it doesn't seem to be Morphia's.

**Comment by Vinarnt** [comment]
It load the long value from database but it was stored in database by encoding with the RoleCodec codec, I expect morphia to load it back using the RoleCodec to decode that long as it was in 1.3.2.
I saw that there's a conversion thing in your recents commits, maybe that's closer to what I'm looking for.

**Comment by evanchooly** [comment]
The type in the database is `Int64` and that will always be loaded using
the `LongCodec`.  The conversions stuff seems to be related but it's not
exactly.

**Comment by evanchooly** [comment]
The type in the database is `Int64` and that will always be loaded using
the `LongCodec`.  The conversions stuff seems to be related but it's not
exactly.

**Comment by Vinarnt** [comment]
Yeah and it's internal.
I also tried to encode the role by writting the discriminator for the role before writing the id, but that leave me with the same result than defining the concrete class as the Role class is not in the discriminator lookup map.

**Comment by Vinarnt** [comment]
One solution would be using life cycle hooks, but I won't be able to see myself in a mirror after that.

https://github.com/Vinarnt/morphia-generic-map-testcase/commit/9114861bbb9b65e18f2c2c541dbc73d465bc62e3

**Comment by evanchooly** [comment]
the problem is that you seem to be trying to do some type conversion there and that's not really how morphia works.

**Comment by Vinarnt** [comment]
I don't have that much choice, I need to persist a data that can't be stored as it is.
So it need to be transformed at some point.

**Comment by evanchooly** [comment]
I'm sorry.  It seems pandemics and releases got in the way.  Are you still fighting this?  it occurs to me, rereading this thread, you could simply write a code for the *Role* type and let *that* change how it's read and written.  then your enclosing POJO can simply rely on Morphia doing what it does.

**Comment by Vinarnt** [comment]
I didn't found a solution with morphia so I moved to kmongo for full control over serialization.
What type of code are you refering to? Isn't what custom codecs are used for?
